FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,87 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(
+						feed: cache.feed
+							.compactMap { ($0 as? ManagedFeedImage) }
+							.map {
+								LocalFeedImage(id: $0.id,
+								               description: $0.imageDescription,
+								               location: $0.location,
+								               url: $0.url)
+							},
+						timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				let managedCache = try ManagedCache.getNewUniqueInstance(context: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = NSOrderedSet(array: feed.map { local in
+					let managed = ManagedFeedImage(context: context)
+					managed.id = local.id
+					managed.imageDescription = local.description
+					managed.location = local.location
+					managed.url = local.url
+					return managed
+				})
+				try context.save()
+				completion(nil)
+			} catch {
+				context.reset()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				try ManagedCache.find(in: context).map(context.delete).map(context.save)
+				completion(nil)
+			} catch {
+				context.reset()
+				completion(error)
+			}
+		}
 	}
 }
+
+public class ManagedCache: NSManagedObject {
+	@NSManaged var timestamp: Date
+	@NSManaged var feed: NSOrderedSet
+
+	static func getNewUniqueInstance(context: NSManagedObjectContext) throws -> ManagedCache {
+		try find(in: context).map(context.delete)
+		return ManagedCache(context: context)
+	}
+
+	static func find(in context: NSManagedObjectContext) throws -> ManagedCache? {
+		guard let name = entity().name else { return nil }
+		let request = NSFetchRequest<ManagedCache>(entityName: name)
+		request.returnsObjectsAsFaults = false
+		return try context.fetch(request).first
+	}
+}
+
+public class ManagedFeedImage: NSManagedObject {
```

Comment:
`ManagedFeedImage` should be `private final`, the same way as `ManagedCache` should.

Please make the class `private final`.

```suggestion
private final class ManagedFeedImage: NSManagedObject {
```